skypy/halos/_colossus.py

```diff
@@ -72,3 +75,131 @@ def colossus_mass_sampler(redshift, model, mdef, m_min, m_max,
     n_uniform = np.random.uniform(size=size)
     masssample = np.interp(n_uniform, CDF, m)
     return masssample
+
+
+@units.quantity_input(sky_area=units.sr)
+def colossus_mf_redshift(redshift, model, mdef, m_min, m_max, sky_area,
+                         cosmology, sigma8, ns, resolution=1000, noise=True):
+    r'''Sample redshifts from a COLOSSUS halo mass function.
+
+    Sample the redshifts of dark matter halos following a mass function
+    implemented in COLOSSUS [1]_ within given mass and redshift ranges and for
+    a given area of the sky.
+
+    Parameters
+    ----------
+    redshift : array_like
+        Input redshift grid on which the mass function is evaluated. Halos are
+        sampled over this redshift range.
+    model : string
+        Mass function model which is available in colossus.
+    mdef : str
+        Halo mass definition for spherical overdensities used by colossus.
+    m_min, m_max : float
+        Lower and upper bounds for the halo mass in units of Solar mass, Msun.
+    sky_area : `~astropy.units.Quantity`
+        Sky area over which halos are sampled. Must be in units of solid angle.
+    cosmology : Cosmology
+        Cosmology object to convert comoving density.
+    sigma8 : float
+        Cosmology parameter, amplitude of the (linear) power spectrum on the
+        scale of 8 Mpc/h.
+    ns : float
+        Cosmology parameter, spectral index of scalar perturbation power
+        spectrum.
+    noise : bool, optional
+        Poisson-sample the number of halos. Default is `True`.
+
+    Returns
+    -------
+    redshifts : array_like
+        Redshifts of the halo sample.
+
+    References
+    ----------
+    .. [1] Diemer B., 2018, ApJS, 239, 35
+
+    '''
+    from colossus.cosmology.cosmology import fromAstropy
+    from colossus.lss.mass_function import massFunction
+
+    # Set the cosmology in COLOSSUS
+    fromAstropy(cosmology, sigma8, ns)
+
+    # Integrate the mass function to get the number density of halos at each redshift
+    def dndlnM(lnm, z):
+        return massFunction(np.exp(lnm), z, 'M', 'dndlnM', mdef, model)
+    lnmmin = np.log(m_min/cosmology.h)
+    lnmmax = np.log(m_max/cosmology.h)
+    density = [integrate.quad(dndlnM, lnmmin, lnmmax, args=(z))[0] for z in redshift]
+    density = np.array(density) * np.power(cosmology.h, 3)
```

Comment:
I am confused about the h units.
The colossus [massFunction](https://bdiemer.bitbucket.io/colossus/lss_mass_function.html#lss.mass_function.massFunction)  takes the mass parameters in units of `M_sun/h`. I understand `skypy.galaxies.redshift.redshifs_from_comoving_density` takes density in units of Mpc^-3.

My first idea was to do `lnmmin = np.log(m_min)` & `lnmmax = np.log(m_max)`, instead. But if I do that, the number of halos (after integrating for the given area of sky) does not agree with `len(z-halo)`

Reply:
All of our function arguments and return value are in units of Msun, while the arguments for colossus massFunction need to be in units of Msun. I realise I messed up the conversion and this should be fixed now.

Reply:
I am not sure of that. I can't make `n_halo` and `len(z_halo)` agree with your new changes.

Reply:
There is a unit test checking exactly that and passing though? `m_min` and `m_max` are in units of M_sun and `massFunction` takes masses in units of M_sun/h so you have to do a change of units somewhere; I chose to do the integral in units of M_sun/h and therefore rescaled the limits by h.

Reply:
Strange. That happens when I run the unit test locally on my computer.